### PR TITLE
Add modules to support Google Cloud DNS

### PR DESCRIPTION
**Pull request by walbert947**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

gcdns_record
gcdns_zone
##### ANSIBLE VERSION

```
ansible 2.2.0 (google_dns 0789f17cf9) last updated 2016/05/18 12:13:30 (GMT -700)
  lib/ansible/modules/core: (detached HEAD 92bf802cb8) last updated 2016/05/18 11:19:41 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD e710dc47fe) last updated 2016/05/18 11:19:51 (GMT -700)
  config file = /home/william/.ansible.cfg
  configured module search path = ['/usr/share/ansible']
```
##### SUMMARY

This is an initial commit for modules that can manipulate Google Cloud DNS. The modules can create and delete zones, as well as records within zones.

A testing playbook is available at https://gist.github.com/walbert947/5731264d0ad0c94fcf36245f5d966c92

**EDIT**: New module_util PR
Both modules rely on a refactored version of the gce module_util. The PR is https://github.com/ansible/ansible/pull/15924

Please review @erjohnso 

P.S. This is my first time contributing to a public open source project. I read and (hopefully) followed the module development guidelines, but if there's anything I'm missing code- or process-wise, please let me know.

### CODE REVIEW

**Comment by walbert947** [comment]
The build is failing with several documentation errors like the following:

```
ERROR: DOCUMENTATION.options.zone.required: expected bool
```

A sample doc snippet:

``` yaml
zone:
    description:
        - The DNS domain name of the zone (e.g., example.com).
        - If both I(zone) and I(zone_id) are specifed, I(zone_id) will be
          used.
    required: true if I(zone_id) is not specified
```

If `required` must be a Boolean, what is the preferred value for options that are conditionally required?

**Edit:** I'm just going to set options with conditional requirements to `required: false`, since that seems to be what other core modules are doing.

**Comment by gregdek** [comment]
Thanks @walbert947 for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by walbert947** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @walbert947 for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/google/gcdns_record.py [hunk 1]
```diff
@@ -0,0 +1,789 @@
...
+    module = AnsibleModule(
+        argument_spec = dict(
+            state                 = dict(default='present', choices=['present', 'absent'], type='str'),
+            record                = dict(required=True, aliases=['name'], type='str'),
+            zone                  = dict(type='str'),
+            zone_id               = dict(type='str'),
+            type                  = dict(required=True, choices=SUPPORTED_RECORD_TYPES, type='str'),
+            values                = dict(type='list'),
+            ttl                   = dict(default=300, type='int'),
+            overwrite             = dict(default=False, choices=BOOLEANS, type='bool'),
```
Comment (resmo) [inline cloud/google/gcdns_record.py]: please remove `choices=BOOLEANS` and keep just `type='bool'`. `choices=BOOLEANS` is deprecated

File: cloud/google/gcdns_record.py [hunk 2]
```diff
@@ -0,0 +1,789 @@
...
+    """Main function"""
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            state                 = dict(default='present', choices=['present', 'absent'], type='str'),
+            record                = dict(required=True, aliases=['name'], type='str'),
+            zone                  = dict(type='str'),
+            zone_id               = dict(type='str'),
+            type                  = dict(required=True, choices=SUPPORTED_RECORD_TYPES, type='str'),
+            values                = dict(type='list'),
```
Comment (resmo) [inline cloud/google/gcdns_record.py]: even though this is a list, you can also use it as a single str item. therefore this works:

```
- gcdns_record:
   record: 'www.example.com'
   zone_id: 'example-com'
   type: CNAME
   values: www.example.com.    # Note the trailing dot
```

But values are in plural, To make it more readable I would recommend to add an alias `aliases=['value']` 

```
- gcdns_record:
   record: 'www.example.com'
   zone_id: 'example-com'
   type: CNAME
   value: www.example.com.    # Note the trailing dot
```

**Comment by resmo** [comment]
@walbert947 would you mind split this PR into 2, one for each module?

**Comment by walbert947** [comment]
@resmo While I can split each module into its own PR since they're technically independent, I would prefer that they remain grouped. Both modules are necessary for full management of Google Cloud DNS via Ansible. Also, my testing playbook assumes that both modules are present.

**Comment by raben2** [comment]
keep me updated if you switch to two pr's i will test the modules

**Comment by walbert947** [comment]
@raben2 Just to be clear, are waiting for a final decision to split into two PR's before you start testing, or are only willing to test the modules if I switch to two PR's?

**Comment by resmo** [comment]
@walbert947 it isn't a policy but a recommendation to have one new module per PR. It makes it usually easier for the reviewers.

I checked the code of your modules and to me, they look already pretty solid, good job! But I have no chance to test them, I would appreciate if @raben2 could do.

As I feel, the modules do not need much (or any) work, there is no need to split into 2 PRs for now.

**Comment by resmo** [comment]
did not test, but code looks solid! LGTM

**Comment by gregdek** [comment]
Thanks again to @walbert947 for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by walbert947** [comment]
Thanks, all! However, the PR in https://github.com/ansible/ansible/pull/15924 **MUST** be merged as well, or these modules will not work.

/cc @erjohnso @resmo @supertom

**Comment by alikins** [comment]
https://github.com/ansible/ansible/pull/15924 is failing tests because of python2.4 incompability. Do these modules need to run on python2.4 nodes?

**Comment by walbert947** [comment]
@alikins No. Libcloud dropped support for Python 2.4 a long time ago, and AFAIK, Ansible's GCE modules have always required Python 2.6 or greater.

**Comment by alikins** [comment]
Shipping it.

**Pull request merged**